Sources/PrivacyDashboard/PrivacyDashboardURLBuilder.swift

```diff
@@ -0,0 +1,106 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+final class PrivacyDashboardURLBuilder {
+
+    enum Constant {
+
+        static let screenKey = "screen"
+        static let breakageScreenKey = "breakageScreen"
+        static let openerKey = "opener"
+        static let categoryKey = "category"
+
+        static let menuScreenKey = "menu"
+        static let dashboardScreenKey = "dashboard"
+
+    }
+
+    enum Configuration {
+
+        case startScreen(entryPoint: PrivacyDashboardEntryPoint, variant: PrivacyDashboardVariant)
+        case segueToScreen(_ screen: Screen, entryPoint: PrivacyDashboardEntryPoint)
+
+        func addScreenParameter(to url: URL) -> URL {
+            var screen: Screen
+            switch self {
+            case .startScreen(let entryPoint, let variant):
+                screen = entryPoint.screen(for: variant)
+            case .segueToScreen(let destinationScreen, _):
+                screen = destinationScreen
+            }
+            return url.appendingParameter(name: Constant.screenKey, value: screen.rawValue)
+        }
+
+        func addBreakageScreenParameterIfNeeded(to url: URL) -> URL {
+            if case .startScreen(_, let variant) = self, let breakageScreen = variant.breakageScreen?.rawValue {
+                return url.appendingParameter(name: Constant.breakageScreenKey, value: breakageScreen)
+            }
+            return url
+        }
+
+        func addCategoryParameterIfNeeded(to url: URL) -> URL {
+            if case .startScreen(let entryPoint, _) = self, case .afterTogglePrompt(let category, _) = entryPoint {
+                return url.appendingParameter(name: Constant.categoryKey, value: category)
+            }
+            return url
+        }
+
+        func addOpenerParameterIfNeeded(to url: URL) -> URL {
+            if case .startScreen(let entryPoint, _) = self, case .toggleReport = entryPoint {
+                return url.appendingParameter(name: Constant.openerKey, value: Constant.menuScreenKey)
+            }
+            if case .segueToScreen(_, let entryPoint) = self, entryPoint == .dashboard {
+                return url.appendingParameter(name: Constant.openerKey, value: Constant.dashboardScreenKey)
+            }
+            return url
+        }
+
+    }
+
+    private var url: URL
+    private let configuration: Configuration
+
+    init(configuration: Configuration) {
+        guard let baseURL = Bundle.privacyDashboardURL else { fatalError() }
+        self.url = baseURL
+        self.configuration = configuration
+    }
+
+    func build() -> URL {
+        url = configuration.addScreenParameter(to: url)
+        url = configuration.addBreakageScreenParameterIfNeeded(to: url)
+        url = configuration.addCategoryParameterIfNeeded(to: url)
+        url = configuration.addOpenerParameterIfNeeded(to: url)
```

Comment:
These functions could be rewritten as URL extension functions taking configuration parameter and then you could do something like this:
```swift
return url.addingScreenParameter(from: configuration)
    .addingBreakageScreenParameterIfNeeded(from: configuration)
    .addingCategoryParameterIfNeeded(from: configuration)
    .addingOpenerParameterIfNeeded(from: configuration)
```

But it really isn't crucial :)

Reply:
I changed this :)